File: lib/js/msgpack-reader.js
"use strict";

module.exports = class Reader {

	constructor(buf, pos) {
		this.buf = buf;
		this.pos = pos | 0;
		this.format = 0;
	}

	reset(buf, pos) {
		this.buf = buf;
		this.pos = pos | 0;
		this.format = 0;
	}

	nextFormat() {
		return this.buf[this.pos];
	}

	readFormat() {
		return this.format = this.buf[this.pos++];
	}

	readBool() {
		return this.format == 0xc3;
	}

	readPositiveFixInt() {
		return this.format & 0x7f;
	}

	readUInt8() {
		return this.buf[this.pos++];
	}

	readUInt16() {
		this.pos += 2;
		return this.buf.readUInt16BE(this.pos - 2);
	}

	readUInt32() {
		this.pos += 4;
		return this.buf.readUInt32BE(this.pos - 4);
	}

	readUInt64() {
		this.pos += 8;
		return this.buf.readUInt32BE(this.pos - 8) * 4294967296 + this.buf.readUInt32BE(this.pos - 4);
	}

	readNegativeFixInt() {
		return (this.format & 0x1f) - 0x20;
	}

	readInt16() {
		this.pos += 2;
		return this.buf.readInt16BE(this.pos - 2);
	}

	readInt32() {
		this.pos += 4;
		return this.buf.readInt32BE(this.pos - 4);
	}

	readInt64() {
		this.pos += 8;
		return this.buf.readInt32BE(this.pos - 8) * 4294967296 + this.buf.readUInt32BE(this.pos - 4);
	}

	readFloat() {
		this.pos += 4;
		return this.buf.readFloatBE(this.pos - 4);
	}

	readDouble() {
		this.pos += 8;
		return this.buf.readDoubleBE(this.pos - 8);
	}

	readFixStr() {
		const length = this.format & 0x1f;
		this.pos += length;
		return this.buf.toString('utf-8', this.pos - length, this.pos);
	}

	readStr8() {
		const length = this.readUInt8();
		this.pos += length;
		return this.buf.toString('utf-8', this.pos - length, this.pos);
	}

	readStr16() {
		const length = this.readUInt16();
		this.pos += length;
		return this.buf.toString('utf-8', this.pos - length, this.pos);
	}

	readStr32() {
		const length = this.readUInt32();
		this.pos += length;
		return this.buf.toString('utf-8', this.pos - length, this.pos);
	}

	readBin8() {
		const length = this.readUInt8();
		this.pos += length;
		return this.buf.slice(this.pos - length, this.pos);
	}

	readBin16() {
		const length = this.readUInt16();
		this.pos += length;
		return this.buf.slice(this.pos - length, this.pos);
	}

	readBin32() {
		const length = this.readUInt32();
		this.pos += length;
		return this.buf.slice(this.pos - length, this.pos);
	}

	readArrayLength() {
		if (this.format == 0xc0) {
			return 0;
		} else if (this.format == 0xdc) {
			return this.readUInt16();
		} else if (this.format == 0xdd) {
			return this.readUInt32();
		} if ((this.format & 0xf0) === 0x90) {
			return this.format & 0xf;
		}
		throw new Error();
	}

	readMapLength() {
		if (this.format == 0xc0) {
			return 0;
		} else if (this.format == 0xde) {
			return this.readUInt16();
		} else if (this.format == 0xdf) {
			return this.readUInt32();
		} if ((this.format & 0xf0) === 0x80) {
			return this.format & 0xf;
		}
		throw new Error();
	}

	readExtLength() {
		switch (this.format) {
			case 0xd4: return 1;
			case 0xd5: return 2;
			case 0xd6: return 4;
			case 0xd7: return 8;
			case 0xd8: return 16;
			case 0xc7: return this.readUInt8();
			case 0xc8: return this.readUInt16();
			case 0xc9: return this.readUInt32();
		}
		throw new Error();
	}

	readExtType() {
		if (this.format == 0xcc) {
			return this.readUInt8();
		} else if (this.format == 0xd0) {
			return this.readUInt8();
		} else if ((this.format & 0x80) === 0x00) {
			return this.readPositiveFixInt();
		} else if ((this.format & 0xe0) === 0xe0) {
			return this.readNegativeFixInt();
		}
		throw new Error();
	}

	readExt(length) {
		this.pos += length;
		return this.buf.slice(this.pos - length, this.pos);
	}

	skip() {
		switch (this.readFormat()) {
			case 0xc0:
			case 0xc2:
			case 0xc3:
				return;
			case 0xcc:
			case 0xd0:
				this.pos += 1;
				return;
			case 0xcd:
			case 0xd1:
				this.pos += 2;
				return;
			case 0xce:
			case 0xd2:
			case 0xca:
				this.pos += 4;
				return;
			case 0xcf:
			case 0xd3:
			case 0xcb:
				this.pos += 8;
				return;
			case 0xd9:
			case 0xc4:
				this.pos += this.readUInt8();
				return;
			case 0xda:
			case 0xc5:
				this.pos += this.readUInt16();
				return;
			case 0xdb:
			case 0xc6:
				this.pos += this.readUInt32();
				return;
			case 0xd4:
				this.pos += 2;
				return;
			case 0xd5:
				this.pos += 3;
				return;
			case 0xd6:
				this.pos += 5;
				return;
			case 0xd7:
				this.pos += 9;
				return;
			case 0xd8:
				this.pos += 17;
				return;
			case 0xc7:
				this.pos += 1 + this.readUInt8();
				return;
			case 0xc8:
				this.pos += 1 + this.readUInt16();
				return;
			case 0xc9:
				this.pos += 1 + this.readUInt32();
				return;
		}
		if ((this.format & 0x80) === 0x00 || (this.format & 0xe0) === 0xe0) {
			return;
		}
		else if ((this.format & 0xe0) === 0xa0) {
			this.pos += this.format & 0x1f;
		}
		else if ((this.format & 0xf0) === 0x90) {
			for (let length = this.readArrayLength(); length > 0; length--) {
				this.skip();
			}
		}
		else if ((this.format & 0xf0) === 0x80) {
			for (let length = this.readMapLength(); length > 0; length--) {
				this.skip();
				this.skip();
			}
		}
	}

}

File: lib/cs/MsgPackReader.cs
using System;
using ILib.MsgPack;

namespace ILib.ProtoPack
{
	public class MsgPackReader : IReader
	{
		static readonly byte[] s_Empty = new byte[0];

		Reader m_Reader;
		public int Offset { get { return m_Reader.CurrentPos; } }

		public MsgPackReader() : this(s_Empty, 0)
		{
		}

		public MsgPackReader(byte[] buf, int offest = 0)
		{
			m_Reader = new Reader(buf, offest);
		}

		public void Reset(byte[] buf, int offest = 0)
		{
			m_Reader.Reset(buf, offest);
		}

		string MissMatchErrorText()
		{
			return string.Format("MsgPackReader fail read format miss match 0x{0:X2}", m_Reader.CurrentFormat);
		}

		public void Skip()
		{
			m_Reader.Skip();
		}

		public uint ReadTag()
		{
			return ReadUint();
		}

		public bool NextFormatIsNull()
		{
			return m_Reader.NextFormat() == Format.Nil;
		}

		public void ReadNil()
		{
			if (m_Reader.ReadFormat() != Format.Nil)
			{
				throw new InvalidOperationException(MissMatchErrorText());
			}
		}

		public T ReadMessage<T>() where T : IMessage, new()
		{
			if (m_Reader.NextFormat() == Format.Nil)
			{
				ReadNil();
				return default(T);
			}
			else
			{
 				var ret = new T();
				ret.Read(this);
				return ret;
			}
		}

		public byte[] ReadBytes()
		{
			if (m_Reader.NextFormat() == Format.Nil)
			{
				ReadNil();
				return null;
			}
			ArraySegment<byte> source;
			switch (m_Reader.ReadFormat())
			{
				case Format.Bin8:
					source = m_Reader.ReadBin8();
					break;
				case Format.Bin16:
					source = m_Reader.ReadBin16();
					break;
				case Format.Bin32:
					source = m_Reader.ReadBin32();
					break;
				default:
					throw new InvalidOperationException(MissMatchErrorText());
			}
			var buf = new byte[source.Count];
			Buffer.BlockCopy(source.Array, source.Offset, buf, 0, source.Count);
			return buf;
		}

		public bool ReadBool()
		{
			switch (m_Reader.ReadFormat())
			{
				case Format.True:
					return true;
				case Format.False:
					return false;
			}
			throw new InvalidOperationException(MissMatchErrorText());
		}

		public int ReadInt()
		{
			return (int)ReadLong();
		}

		public uint ReadUint()
		{
			return (uint)ReadUlong();
		}

		public long ReadLong()
		{
			var format = m_Reader.ReadFormat();
			switch (format)
			{
				case Format.Float64:
					return (long)m_Reader.ReadFloat64();
				case Format.Float32:
					return (long)m_Reader.ReadFloat32();
				case Format.UInt8:
					return m_Reader.ReadUInt8();
				case Format.UInt16:
					return m_Reader.ReadUInt16();
				case Format.UInt32:
					return m_Reader.ReadUInt32();
				case Format.UInt64:
					return (long)m_Reader.ReadUInt64();
				case Format.Int8:
					return m_Reader.ReadInt8();
				case Format.Int16:
					return m_Reader.ReadInt16();
				case Format.Int32:
					return m_Reader.ReadInt32();
				case Format.Int64:
					return m_Reader.ReadInt64();
			}
			if (Format.IsPositiveFixInt(format))
			{
				return m_Reader.ReadPositiveFixInt();
			}
			else if (Format.IsNegativeFixInt(format))
			{
				return m_Reader.ReadNegativeFixInt();
			}
			throw new InvalidOperationException(MissMatchErrorText());
		}

		public ulong ReadUlong()
		{
			var format = m_Reader.ReadFormat();
			switch (format)
			{
				case Format.UInt8:
					return m_Reader.ReadUInt8();
				case Format.UInt16:
					return m_Reader.ReadUInt16();
				case Format.UInt32:
					return m_Reader.ReadUInt32();
				case Format.UInt64:
					return m_Reader.ReadUInt64();
				case Format.Int8:
					return (ulong)m_Reader.ReadInt8();
				case Format.Int16:
					return (ulong)m_Reader.ReadInt16();
				case Format.Int32:
					return (ulong)m_Reader.ReadInt32();
				case Format.Int64:
					return (ulong)m_Reader.ReadInt64();
			}
			if (Format.IsPositiveFixInt(format))
			{
				return m_Reader.ReadPositiveFixInt();
			}
			throw new InvalidOperationException(MissMatchErrorText());
		}

		public double ReadDouble()
		{
			if (m_Reader.NextFormat() == Format.Float64)
			{
				m_Reader.ReadFormat();
				return m_Reader.ReadFloat64();
			}
			else if (m_Reader.NextFormat() == Format.Float32)
			{
				m_Reader.ReadFormat();
				return m_Reader.ReadFloat32();
			}
			return ReadLong();
		}

		public float ReadFloat()
		{
			if(m_Reader.NextFormat() == Format.Float32) {
				m_Reader.ReadFormat();
				return m_Reader.ReadFloat32();
			}
			return ReadLong();
		}

		public string ReadString()
		{
			var format = m_Reader.ReadFormat();
			switch (format)
			{
				case Format.Str8:
					return m_Reader.ReadStr8();
				case Format.Str16:
					return m_Reader.ReadStr16();
				case Format.Str32:
					return m_Reader.ReadStr32();
				case Format.Nil:
					return null;
			}
			if(Format.IsFixStr(format))
			{
				return m_Reader.ReadFixStr();
			}
			throw new InvalidOperationException(MissMatchErrorText());
		}

		public int ReadMapHeader()
		{
			m_Reader.ReadFormat();
			return m_Reader.ReadMapLength();
		}

		public int ReadArrayHeader()
		{
			m_Reader.ReadFormat();
			return m_Reader.ReadArrayLength();
		}
	}
}
